### PR TITLE
HA config implementation based on proposal in PR 1730

### PR DESCRIPTION
**Pull request by ramr**
Initial cut of the HA config implementation based on the proposal in https://github.com/openshift/origin/pull/1730

Example usage: 
```  
#  label some nodes router=us-west-ha
openshift kube label node minion-1 "router=us-west-ha"  
openshift kube label node minion-2 "router=us-west-ha"  

#  create router and ha sidecar on the labelled nodes.  
openshift ex router --create --labels="router=us-west-ha" --credentials="${KUBECONFIG}" --replicas=2  
openshift ex ha-config har --virtual-ips=10.0.2.100-102 --credentials=$KUBECONFIG  --replicas=2 -w 80 --latest-images --selector="router=us-west-ha" --create  
```

FYI, @smarterclayton @pweil-   PTAL and  let me know if this approach seems okay. 

This PR is dependent on the proposal PR - so don't merge till https://github.com/openshift/origin/pull/1730  is merged in. 
So there's probably going to be duplicate cruft  in here (the cleaned up version of the keepalived demo pieces).  

Also there is more work to do to handle nodes joining/leaving the cluster and on their labels changing.

### CODE REVIEW

File: images/ha-config/keepalived/tests/echoserver.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
+#!/usr/bin/env python
+
```
Comment (smarterclayton) [inline images/ha-config/keepalived/tests/echoserver.py]: Would be better to do a go echo server here

File: pkg/haconfig/interfaces.go [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package haconfig
```
Comment (smarterclayton) [inline pkg/haconfig/interfaces.go]: Call this package ipfailover

File: plugins/haconfig/keepalived/helpers.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
+package keepalived
```
Comment (smarterclayton) [inline plugins/haconfig/keepalived/helpers.go]: plugins/ is an artifact of some older decisions, make this pkg/ipfailover/keepalived/...

File: images/ha-config/keepalived/tests/verify_failover_image.sh [hunk 4]
```diff
@@ -0,0 +1,90 @@
+#!/bin/bash -e
```
Comment (smarterclayton) [inline images/ha-config/keepalived/tests/verify_failover_image.sh]: Would be better as an integration test like the routers

File: plugins/haconfig/keepalived/generator.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+	if c != nil {
+		containers = append(containers, *c)
+	}
+
+	return containers
+}
+
+//  Generate the HA failover monitor (keepalived) container volume config.
+func generateVolumeConfig() []kapi.Volume {
+	hostPath := &kapi.HostPathVolumeSource{Path: libModulesPath}
```
Comment (smarterclayton) [inline plugins/haconfig/keepalived/generator.go]: Why is this necessary?

File: images/ha-config/keepalived/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,17 @@
+#
+# VIP failover monitoring container for OpenShift Origin.
+#
+# ImageName: openshift/origin-keepalived-ha-config
+#
+# FROM fedora
```
Comment (smarterclayton) [inline images/ha-config/keepalived/Dockerfile]: Remove

File: pkg/cmd/experimental/haconfig/haconfig.go [hunk 7]
```diff
@@ -0,0 +1,188 @@
+package haconfig
```
Comment (smarterclayton) [inline pkg/cmd/experimental/haconfig/haconfig.go]: ipfailover

File: images/ha-config/keepalived/lib/config-generators.sh [hunk 8]
```diff
@@ -0,0 +1,268 @@
+#!/bin/bash
```
Comment (smarterclayton) [inline images/ha-config/keepalived/lib/config-generators.sh]: Why bash and not a go template like router?

File: plugins/haconfig/keepalived/helpers.go [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+)
+
+type ReporterFunction func(string, error) 
+
+
+//  Delete matching replication controllers.
+func deleteReplicationControllers(name string, namespace string, selector map[string]string, kClient kclient.Interface, after ReporterFunction) {
+	labels := labels.SelectorFromSet(selector)
```
Comment (smarterclayton) [inline plugins/haconfig/keepalived/helpers.go]: Why is this needed?

File: plugins/haconfig/keepalived/plugin.go [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+	"github.com/golang/glog"
+
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	configcmd "github.com/openshift/origin/pkg/config/cmd"
+	"github.com/openshift/origin/pkg/generate/app"
+	"github.com/openshift/origin/pkg/haconfig"
+)
+
+//  HA configurator plugin for keepalived sidecar.
+type KeepAlivedPlugin struct {
```
Comment (smarterclayton) [inline plugins/haconfig/keepalived/plugin.go]: Nit: it's probably easier to keep your sanity killing it KeepalivedPlugin

Comment (smarterclayton) [inline plugins/haconfig/keepalived/plugin.go]: Calling it

File: docs/proposals/ha-configuration.md [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+watched service (e.g. HAProxy router) - not all nodes would be allocated a
+virtual IP but would still be candidates on a failure.
+And the case where the number of VIPs is more than the node selection, a
+node could service multiple VIPs.
+
+This allows us to the cover:
+  1. The normal steady state case.
+  1. When a cluster is resized - nodes shrink or grow.
+  1. When a cluster is modified - really node labels are modified.
+  1. Failure cases when a node or watched service or network fails.
```
Comment (smarterclayton) [inline docs/proposals/ha-configuration.md]: Add a note here that the PerNodeController will remove the need to fiddle with this, because you will be able to run these on a set of nodes directly.

File: pkg/cmd/experimental/haconfig/haconfig.go [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+
+  Create a HA configuration on a selection of nodes labelled
+  "router=us-west-ha" (on 4 nodes with 7 virtual IPs monitoring a service
+  listening on port 80 (aka the OpenShift router process).
+
+  $ %[1]s %[2]s ha-config --selector="router=us-west-ha" --virtual-ips="1.2.3.4,10.1.1.100-104,5.6.7.8" --watch-port=80 --replicas=4 --create
+
+  Delete a previously created HA configuration:
+
+  $ %[1]s %[2]s hac --delete
```
Comment (smarterclayton) [inline pkg/cmd/experimental/haconfig/haconfig.go]: Don't need this, we have the "stop" command and delete for this.  Just ensure you have a default label setup.

File: docs/proposals/ha-configuration.md [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+     required to perform any manual interaction to update the upstream
+     "discovery" sources (e.g. DNS). The cluster should service all the
+     assigned virtual IPs when atleast a single node is available - and
+     this should be inspite of the fact that the current available
+     resources are not sufficient to reach "critical mass" aka the
+     desired state.
+
+
+## Goals
+The goal here is to provide the OpenShift environment with one or more
```
Comment (smarterclayton) [inline docs/proposals/ha-configuration.md]: Solution is specific to node ips.  What would it take to give pods stable VIPs?  Can the ipfailover pod handle VIPs for sets of pod ips?  What changes would be required to support both and how much work would it be?

How does Neil's external ip solution affect this setup?  If we have routers an external ip, would this be able to manage that?

File: plugins/haconfig/keepalived/generator.go [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+//  Generate the HA failover monitor (keepalived) container configuration.
+func generateFailoverMonitorContainerConfig(name string, options *haconfig.HAConfigCmdOptions, env app.Environment) *kapi.Container {
+	containerName := fmt.Sprintf("%s-%s", name, options.Type)
+
+	imageName := fmt.Sprintf("%s-%s", options.Type, haconfig.DefaultName)
+	image := options.ImageTemplate.ExpandOrDie(imageName)
+
+	ports := make([]kapi.ContainerPort, 1)
+	ports[0] = kapi.ContainerPort{
+		ContainerPort: options.ServicePort,
```
Comment (smarterclayton) [inline plugins/haconfig/keepalived/generator.go]: Is this keepaliveds interconnect port?  Is that their default assigned port?  If so, add a comment to that effect

File: plugins/haconfig/keepalived/generator.go [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+	return dapi.DeploymentTemplate{
+		Strategy: dapi.DeploymentStrategy{
+			Type: dapi.DeploymentStrategyTypeRecreate,
+		},
+		ControllerTemplate: kapi.ReplicationControllerSpec{
+			// TODO: v0.1 requires a manual resize to the
+			//       replicas to match current cluster state.
+			//       v0.1+ could do this with either a watcher
+			//       that updates the replica count or better
+			//       yet, some way to kubernetes to say run
```
Comment (smarterclayton) [inline plugins/haconfig/keepalived/generator.go]: That's what PerNodeController will do

File: plugins/haconfig/keepalived/generator.go [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+
+	return &dapi.DeploymentConfig{
+		ObjectMeta: kapi.ObjectMeta{
+			Name:   name,
+			Labels: selector,
+		},
+		Triggers: []dapi.DeploymentTriggerPolicy{
+			{Type: dapi.DeploymentTriggerOnConfigChange},
+		},
+		Template: generateDeploymentTemplate(name, options, selector),
```
Comment (smarterclayton) [inline plugins/haconfig/keepalived/generator.go]: Doesn't seem to make it any more readable to break it out - recommend just adding the template here.

File: images/ha-config/keepalived/tests/echoserver.py (hunk 1 above)
Comment (ramr) [inline images/ha-config/keepalived/tests/echoserver.py]: Yeah, this was just to test and verify the image is okay before I had the integration code in place. The image verification test monitors the echo server listen port.

File: pkg/haconfig/interfaces.go (hunk 2 above)
Comment (ramr) [inline pkg/haconfig/interfaces.go]: Will do and change it across the board.

File: plugins/haconfig/keepalived/helpers.go (hunk 3 above)
Comment (ramr) [inline plugins/haconfig/keepalived/helpers.go]: Hmm, ok - will do.

File: pkg/haconfig/interfaces.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/haconfig/interfaces.go]: I think ipfailover is strictly better than haconfig, but if you have a different suggestion let me know.  Other options are vipmanager or vipfailover - I didn't brainstorm extensively.

----- Original Message -----

File: images/ha-config/keepalived/tests/echoserver.py (hunk 1 above)
Comment (smarterclayton) [inline images/ha-config/keepalived/tests/echoserver.py]: K, just noticed it while i was going through.

----- Original Message -----

File: images/ha-config/keepalived/tests/verify_failover_image.sh (hunk 4 above)
Comment (ramr) [inline images/ha-config/keepalived/tests/verify_failover_image.sh]: How about I add a separate ipfailover integration test. This one is really a docker image verification test, so might be good to keep it around.

File: plugins/haconfig/keepalived/generator.go (hunk 5 above)
Comment (ramr) [inline plugins/haconfig/keepalived/generator.go]: The keepalived sidecar container needs access to the kernel modules directory in order to load/unload it.

File: images/ha-config/keepalived/Dockerfile (hunk 6 above)
Comment (ramr) [inline images/ha-config/keepalived/Dockerfile]: Will do.

File: pkg/cmd/experimental/haconfig/haconfig.go (hunk 7 above)
Comment (ramr) [inline pkg/cmd/experimental/haconfig/haconfig.go]: Will do.

File: plugins/haconfig/keepalived/generator.go (hunk 5 above)
Comment (smarterclayton) [inline plugins/haconfig/keepalived/generator.go]: Ok, add a comment here and make it clear this is why it's needed.

----- Original Message -----

File: images/ha-config/keepalived/lib/config-generators.sh (hunk 8 above)
Comment (ramr) [inline images/ha-config/keepalived/lib/config-generators.sh]: Was what I had from the original demo but it could be anything. Although that said, it does simplify the innards of the sidecar pod to have bash script. I can rework this in a subsequent PR if needed.

File: images/ha-config/keepalived/tests/verify_failover_image.sh (hunk 4 above)
Comment (smarterclayton) [inline images/ha-config/keepalived/tests/verify_failover_image.sh]: Sure, I just want to keep tests focused.

----- Original Message -----

File: plugins/haconfig/keepalived/helpers.go (hunk 9 above)
Comment (ramr) [inline plugins/haconfig/keepalived/helpers.go]: So I had a conversation with @abhgupta yesterday and was asking re: the cleanup/deletion of these objects and how to go about doing it. The ```--delete``` option cleans up the deployment artifacts and uses this.

File: images/ha-config/keepalived/lib/config-generators.sh (hunk 8 above)
Comment (smarterclayton) [inline images/ha-config/keepalived/lib/config-generators.sh]: It's fine as a follow on, leave a TODO.  Doesn't seem like anything bash specific is needed, just simple templatization.

----- Original Message -----

File: plugins/haconfig/keepalived/helpers.go (hunk 9 above)
Comment (smarterclayton) [inline plugins/haconfig/keepalived/helpers.go]: We generally just don't do that from this code base - that's why we have stop and delete and others (other ways to do this that are consistent).

----- Original Message -----

File: plugins/haconfig/keepalived/plugin.go (hunk 10 above)
Comment (ramr) [inline plugins/haconfig/keepalived/plugin.go]: Will change that - yeah, was debating over that spelling as well - changed it back and forth.

File: pkg/cmd/experimental/haconfig/haconfig.go (hunk 12 above)
Comment (ramr) [inline pkg/cmd/experimental/haconfig/haconfig.go]: Ok, will remove this out. The default label setup is already there.

File: docs/proposals/ha-configuration.md (hunk 11 above)
Comment (ramr) [inline docs/proposals/ha-configuration.md]: Will do.

File: plugins/haconfig/keepalived/generator.go (hunk 14 above)
Comment (ramr) [inline plugins/haconfig/keepalived/generator.go]: Will do.

File: pkg/haconfig/interfaces.go (hunk 2 above)
Comment (ramr) [inline pkg/haconfig/interfaces.go]: Yeah, I didn't have a strong preference/suggestion for the name. 
An ipfailover by any other name s(m)ells just as sweet.

File: plugins/haconfig/keepalived/generator.go (hunk 5 above)
Comment (ramr) [inline plugins/haconfig/keepalived/generator.go]: Will do.

File: plugins/haconfig/keepalived/helpers.go (hunk 9 above)
Comment (ramr) [inline plugins/haconfig/keepalived/helpers.go]: Ok - I will remove this and leave a couple of comments/TODOs in place.

File: plugins/haconfig/keepalived/generator.go (hunk 16 above)
Comment (ramr) [inline plugins/haconfig/keepalived/generator.go]: Was just easier for me to follow + readable to break it out into smaller bits and work on it in small chunks. That was just a personal preference - will move it back in.

File: docs/proposals/ha-configuration.md (hunk 13 above)
Comment (ramr) [inline docs/proposals/ha-configuration.md]: As re: this solution - its true in that it pertinent to the network interface that is being managed - in this case the host one - not really the individual pods. If we want to do it at a pod level, we would need some sort of isolation mechanism (```iptables rules```) to block out other pods - Neil's solution would work well here for the isolation component. But we would also need to handle ```multi-tenancy``` for multiple configs on different pods - plus tie-in the two configuration pieces in the pod spec and the ipfailover config together. Needs some more thought on that.

File: plugins/haconfig/keepalived/plugin.go (hunk 10 above)
Comment (smarterclayton) [inline plugins/haconfig/keepalived/plugin.go]: It's ugly both ways but this way you don't have the A poking you in the eye.

----- Original Message -----

File: docs/proposals/ha-configuration.md (hunk 13 above)
Comment (smarterclayton) [inline docs/proposals/ha-configuration.md]: Ok, I just want to know how this could evolve in the long run.  A router might not be scheduled on the current host for some reason, so being able to send traffic from the host to a pod seems similar enough that I'd want to understand why the host is superior.

----- Original Message -----

**Comment by ramr** [comment]
made all fixes as per review comments (and rebased) @smarterclayton PTAL. Thx

File: pkg/cmd/experimental/ipfailover/ipfailover.go [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+Configure or view IP Failover configuration
+
+This command helps to setup IP Failover configuration for an OpenShift
+environment. An administrator can configure IP failover on an entire
+cluster or as would normally be the case on a subset of nodes (as defined
+via a labelled selector).
+
+If an IP failover configuration does not exist with the given name,
+the --create flag can be passed to create a deployment configuration and
+service that will provide IP failover capability. If you are running in
```
Comment (smarterclayton) [inline pkg/cmd/experimental/ipfailover/ipfailover.go]: You don't need a service - or at least, I can't think of a reason you would need one.

File: pkg/cmd/experimental/ipfailover/ipfailover.go [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+
+	cmd.Flags().StringVar(&options.VirtualIPs, "virtual-ips", "", "A set of virtual IP ranges and/or addresses that the routers bind and serve on and provide IP failover capability for.")
+	cmd.Flags().StringVarP(&options.NetworkInterface, "interface", "i", "", "Network interface bound by VRRP to use for the set of virtual IP ranges/addresses specified.")
+
+	// unicastHelp := `Send VRRP adverts using unicast instead of over the VRRP multicast group. This is useful in environments where multicast is not supported. Use with caution as this can get slow if the list of peers is large - it is recommended running this with the label option to select a set of nodes.`
+	// cmd.Flags().StringVarP(&options.UseUnicast, "unicast", "u", options.UseUnicast, unicastHelp)
+
+	cmd.Flags().IntVarP(&options.WatchPort, "watch-port", "w", ipf.DefaultWatchPort, "Port to monitor or watch for resource availability.")
+	cmd.Flags().IntVarP(&options.Replicas, "replicas", "r", options.Replicas, "The replication factor of this IP failover configuration; commonly 2 when high availability is desired.")
```
Comment (smarterclayton) [inline pkg/cmd/experimental/ipfailover/ipfailover.go]: I don't think this comment is accurate - you probably should be telling people to set this number equal to the number of nodes for selector

File: pkg/cmd/experimental/ipfailover/ipfailover.go [hunk 19]
```diff
@@ -0,0 +1,173 @@
...
+		Long:  fmt.Sprintf(description, parentName, name),
+		Run: func(cmd *cobra.Command, args []string) {
+			processCommand(f, options, cmd, args, out)
+		},
+	}
+
+	cmd.Flags().StringVar(&options.Type, "type", ipf.DefaultType, "The type of IP failover configurator to use.")
+	cmd.Flags().StringVar(&options.ImageTemplate.Format, "images", options.ImageTemplate.Format, "The image to base this IP failover configurator on - ${component} will be replaced based on --type.")
+	cmd.Flags().BoolVar(&options.ImageTemplate.Latest, "latest-images", options.ImageTemplate.Latest, "If true, attempt to use the latest images instead of the current release")
+	cmd.Flags().StringVarP(&options.Selector, "selector", "l", options.Selector, "Selector (label query) to filter nodes on.")
```
Comment (smarterclayton) [inline pkg/cmd/experimental/ipfailover/ipfailover.go]: This is required, correct?

File: pkg/cmd/experimental/ipfailover/ipfailover.go [hunk 20]
```diff
@@ -0,0 +1,173 @@
...
+		WatchPort:        ipf.DefaultWatchPort,
+		NetworkInterface: ipf.DefaultInterface,
+		Replicas:         1,
+	}
+
+	cmd := &cobra.Command{
+		Use:   fmt.Sprintf("%s [<name>]", name),
+		Short: shortDesc,
+		Long:  fmt.Sprintf(description, parentName, name),
+		Run: func(cmd *cobra.Command, args []string) {
```
Comment (smarterclayton) [inline pkg/cmd/experimental/ipfailover/ipfailover.go]: This command needs to follow the pattern for the rest of our CLI commands - a RunProcess method that returns an error with cmdutil.CheckErr(err).  Also, you should always be returning errors from functions and glog fatal is reserved for a very few specific locations.

File: pkg/cmd/experimental/ipfailover/ipfailover.go [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+	"fmt"
+	"io"
+
+	cmdutil "github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/cmd/util"
+	"github.com/golang/glog"
+	"github.com/spf13/cobra"
+
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	"github.com/openshift/origin/pkg/cmd/util/variable"
+	ipf "github.com/openshift/origin/pkg/ipfailover"
```
Comment (smarterclayton) [inline pkg/cmd/experimental/ipfailover/ipfailover.go]: Don't use short package aliases in general in Go

File: pkg/ipfailover/keepalived/generator.go [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+	"github.com/openshift/origin/pkg/ipfailover"
+)
+
+const defaultInterface = "eth0"
+const libModulesVolumeName = "lib-modules"
+const libModulesPath = "/lib/modules"
+
+//  Get kube client configuration from a file containing credentials for
+//  connecting to the master.
+func getClientConfig(path string) *kclient.Config {
```
Comment (smarterclayton) [inline pkg/ipfailover/keepalived/generator.go]: Return errors, don't use glog.Fatal

File: pkg/ipfailover/keepalived/plugin.go [hunk 23]
```diff
@@ -0,0 +1,130 @@
...
+	}
+
+	namespace := p.GetNamespace()
+	service, err := kClient.Services(namespace).Get(p.Name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			glog.V(4).Infof("KeepAlived IP Failover config: %s - no service found", p.Name)
+			return nil
+		}
+		glog.Fatalf("Error getting KeepAlived IP Failover config service %q: %v", p.Name, err)
```
Comment (smarterclayton) [inline pkg/ipfailover/keepalived/plugin.go]: return error

File: pkg/ipfailover/keepalived/plugin.go [hunk 24]
```diff
@@ -0,0 +1,130 @@
...
+func (p *KeepalivedPlugin) Create(out io.Writer) {
+	namespace := p.GetNamespace()
+
+	bulk := configcmd.Bulk{
+		Factory: p.Factory.Factory,
+		After:   configcmd.NewPrintNameOrErrorAfter(out, os.Stderr),
+	}
+
+	if errs := bulk.Create(p.Generate(), namespace); len(errs) != 0 {
+		glog.Fatalf("Error creating config: %+v", errs)
```
Comment (smarterclayton) [inline pkg/ipfailover/keepalived/plugin.go]: return error

File: pkg/ipfailover/types.go [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+	Create        bool
+
+	//  Failover options.
+	VirtualIPs       string
+	NetworkInterface string
+	WatchPort        int
+	Replicas         int
+
+	//  For the future - currently unused.
+	UseUnicast bool
```
Comment (smarterclayton) [inline pkg/ipfailover/types.go]: Remove it if it's unused.

File: pkg/cmd/experimental/ipfailover/ipfailover.go (hunk 17 above)
Comment (ramr) [inline pkg/cmd/experimental/ipfailover/ipfailover.go]: Ok - will remove it.

File: pkg/cmd/experimental/ipfailover/ipfailover.go (hunk 18 above)
Comment (ramr) [inline pkg/cmd/experimental/ipfailover/ipfailover.go]: Will update it.

File: pkg/cmd/experimental/ipfailover/ipfailover.go (hunk 19 above)
Comment (ramr) [inline pkg/cmd/experimental/ipfailover/ipfailover.go]: There is a default selector if one is not passed.

File: pkg/cmd/experimental/ipfailover/ipfailover.go (hunk 20 above)
Comment (ramr) [inline pkg/cmd/experimental/ipfailover/ipfailover.go]: ok - will do the error propogation up and check for it.

File: pkg/cmd/experimental/ipfailover/ipfailover.go (hunk 21 above)
Comment (ramr) [inline pkg/cmd/experimental/ipfailover/ipfailover.go]: Ok - will do.

File: pkg/ipfailover/keepalived/generator.go (hunk 22 above)
Comment (ramr) [inline pkg/ipfailover/keepalived/generator.go]: Yup - will bubble the errors all the way up.

**Comment by ramr** [comment]
@smarterclayton made fixes as per your review comments.

**Comment by smarterclayton** [comment]
LGTM [merge] - we can track other items in later pulls.  Thanks
